arxiv/base/urls/tests/test_urlize.py

```diff
@@ -197,47 +197,77 @@ def test_parens(self):
-            'images is at <a class="link-external link-http" href="http://85.20.11.14/hosting/punsly/APJLetter4.2.07/" rel="external">this http URL</a>',
+            'images is at <a class="link-external link-http" href="http://85.20.11.14/hosting/punsly/APJLetter4.2.07/" rel="external noopener nofollow">this http URL</a>',
             "should deal with numeric IP correctly"
         )
 
     @mock.patch(f'{links.__name__}.url_for', mock_url_for)
     def test_urls_with_plus(self):
         self.assertEqual(
             links.urlize('http://www.fkf.mpg.de/andersen/docs/pub/abstract2004+/pavarini_02.pdf'),
-            '<a class="link-external link-http" href="http://www.fkf.mpg.de/andersen/docs/pub/abstract2004+/pavarini_02.pdf" rel="external">this http URL</a>'
+            '<a class="link-external link-http" href="http://www.fkf.mpg.de/andersen/docs/pub/abstract2004+/pavarini_02.pdf" rel="external noopener nofollow">this http URL</a>'
         )
 
     @mock.patch(f'{links.__name__}.url_for', mock_url_for)
     def test_anchors_with_slash(self):
-        self.assertEqual(
+        self.assertIn(
+            'href="https://dms.sztaki.hu/ecml-pkkd-2016/#/app/privateleaderboard"',
             links.urlize('https://dms.sztaki.hu/ecml-pkkd-2016/#/app/privateleaderboard'),
-            '<a class="link-external link-https" href="https://dms.sztaki.hu/ecml-pkkd-2016/#/app/privateleaderboard" rel="external">this https URL</a>',
             "Should deal with slash in URL anchor correctly"
         )
 
     @mock.patch(f'{links.__name__}.url_for', mock_url_for)
     def test_ftp(self):
         self.assertEqual(
             links.urlize('7 Pages; ftp://ftp%40micrognu%2Ecom:anon%40anon@ftp.micrognu.com/pnenp/conclusion.pdf'),
-            '7 Pages; <a class="link-external link-ftp" href="ftp://ftp%40micrognu%2Ecom:anon%40anon@ftp.micrognu.com/pnenp/conclusion.pdf" rel="external">this ftp URL</a>'
+            '7 Pages; <a class="link-external link-ftp" href="ftp://ftp%40micrognu%2Ecom:anon%40anon@ftp.micrognu.com/pnenp/conclusion.pdf" rel="external noopener nofollow">this ftp URL</a>'
        )
 
     @mock.patch(f'{links.__name__}.url_for', mock_url_for)
     def test_arxiv_prefix(self):
         self.assertEqual(
             links.urlize("see arxiv:1201.12345"),
-            'see <a class="link-https" data-arxiv-id="1201.12345" href="https://arxiv.org/abs/1201.12345">arXiv:1201.12345</a>'
-        )
+            'see <a class="link-https" data-arxiv-id="1201.12345" href="https://arxiv.org/abs/1201.12345">arXiv:1201.12345</a>')
+
+
+    @mock.patch(f'{links.__name__}.clickthrough')
+    def test_doi_2(self, mock_clickthrough):
+        mock_clickthrough.clickthrough_url = lambda x: x
+        self.assertRegex(links.urlize('10.1088/1475-7516/2018/07/009'),
+                         r'<a.*href="https://.*10.1088/1475-7516/2018/07/009".*>10.1088/1475-7516/2018/07/009</a>')
```

Comment:
I'm curious about the wildcards here:

`r'<a.*href="https://.*10.1088/1475-7516/2018/07/009".*>10.1088/1475-7516/2018/07/009</a>')`

First wildcard is intended to match against the `class` (and any other) attribute?
Last wildcard is intended to match against `rel` (and any other) attribute?
Is the one after the protocol meant to match against clickthrough linking?

Reply:
In this test I don't really care about the class or rel. This is just testing the href and the text.
So I'm not including those in what gets tested by throwing in the wildcards.

The `mock_clickthrough.clickthrough_url = lambda x: x` part sets the clickthrough function to the identity function because I'm not interested in testing the CT in this test either.

The wildcard after the protocol was meant to skip dx.doi.org.

Reply:
👍